android/app/src/main/java/org/fossasia/openevent/core/schedule/DayScheduleFragmentViewModel.java

```diff
@@ -0,0 +1,88 @@
+package org.fossasia.openevent.core.schedule;
+
+import android.arch.lifecycle.LiveData;
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.lifecycle.ViewModel;
+
+import org.fossasia.openevent.common.arch.MutableFilterableRealmLiveData;
+import org.fossasia.openevent.data.Session;
+import org.fossasia.openevent.data.repository.RealmDataRepository;
+import org.fossasia.openevent.common.utils.SortOrder;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+import io.reactivex.functions.Predicate;
+import io.realm.RealmResults;
+import io.realm.Sort;
+import io.realm.internal.Collection;
+
+public class DayScheduleFragmentViewModel extends ViewModel {
+
+    private RealmDataRepository realmRepo;
+    private MutableFilterableRealmLiveData<Session> mutableFilterSessions;
+    private MutableLiveData<String> sortTypeSchedule;
+
+    private String searchText = "";
+
+    public DayScheduleFragmentViewModel() {
+        realmRepo = RealmDataRepository.getDefaultInstance();
+    }
+
+    public String getSearchText() { return searchText; }
+
+    public void setSearchText(String searchText) {
+        this.searchText = searchText;
+    }
+
+    public LiveData<List<Session>> getSessionsByDate(String date, String searchText) {
+        ensureSessionsAreLoaded(date);
+        filterWithSearch(searchText);
+        sortSessions();
+        return mutableFilterSessions;
+    }
+
+    public LiveData<List<Session>> getSessionsByTracks(String date, String searchText, List<String> selectedTracks) {
+        ensureSessionsAreLoaded(date);
+        filterWithTracks(selectedTracks);
+        filterWithSearch(searchText);
+        sortSessions();
+        return mutableFilterSessions;
+    }
+    private void filterWithSearch(String searchText) {
+        if (!this.searchText.equals(searchText)) {
+            setSearchText(searchText);
+            final String query = searchText.toLowerCase(Locale.getDefault());
+            Predicate<Session> predicate = session -> session.getTitle()
+                    .toLowerCase(Locale.getDefault())
+                    .contains(query);
+            mutableFilterSessions.filter(predicate);
+        }
+    }
+
+    private void filterWithTracks(List<String> selectedTracks) {
+        if (!selectedTracks.isEmpty()) {
+            Predicate<Session> predicate = session -> selectedTracks.contains(session.getTrack().getName());
+            mutableFilterSessions.filter(predicate);
+        }
+    }
+
+    private void sortSessions() {
+        List<Session> sessionList = mutableFilterSessions.getValue();
+
+        // Sort sessions according to global sortTypeSchedule
+        if (SortOrder.sortOrderSchedule() == SortOrder.SORT_ORDER_DESCENDING) {
+            sessionList = ((RealmResults<Session>) sessionList).sort(SortOrder.sortTypeSchedule(), Sort.DESCENDING);
```

Comment:
The only issue now is that the app crashes when filtering with tracks. The error is at line 76, java.util.ArrayList sessionList cannot be cast to io.realm.RealmResults. Is there an easier way of sorting by SortOrder.sortTypeSchedule() instead of casting to RealmResults?

Reply:
@leeyh20 Yes you cannot cast ArrayList to RealmResults. This works fine until `filterWithTracks` is executed. In this method, this line -> `mutableFilterSessions.filter(predicate);` changes the type of value stored in mutableFilterSessions from **RealmResults** to **ArrayList** and hence when you execute `mutableFilterSessions.getValue()` after this, you get a return type of ArrayList.
Check the super class method which you are using here.

Reply:
@mayank8318 Thanks! I’ll try to solve this.